### PR TITLE
Documentation: transaction dev guide v5

### PR DESCRIPTION
**Pull request by jufajardini**
Describe changes:
- Add contents and examples about rule matching
- Fix typos
- Change formatting

Link to Redmine issue: https://redmine.openinfosecfoundation.org/issues/4396

Previous PR: https://github.com/OISF/suricata/pull/6068

### CODE REVIEW

**Comment by jufajardini** [comment]
Must still add Sequence Diagrams.

File: doc/devguide/extending/app-layer/transactions.rst [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+In addition, for file transfer protocols, or similar ones where there may be several messages before the file exchange
+is completed (NFS, SMB), it is possible to create a level of abstraction to handle such complexity. This could be achieved by adding phases to the protocol implemented model (e.g., protocol negotiation phase (SMB), request parsed (HTTP), and so on).
+
+This is controlled by implementing states. In Suricata, those will be enums that are incremented as the parsing
+progresses. A state will start at 0. The higher its value, the closer the transaction would be to completion.
+
+The engine interacts with transactions state using a set of callbacks the parser registers. State is defined per flow direction (``STREAM_TOSERVER`` / ``STREAM_TOCLIENT``).
+
+In Summary - Transactions and State
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:120]: nit -- extends further than text in L119

File: doc/devguide/extending/app-layer/transactions.rst [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+- Initial state value: ``0``
+- Simpler scenarios: state is simply an int.  ``1`` represents transaction completion, per direction.
+- Complex Transaction State in Suricata: ``enum`` (Rust: ``i32``). Completion is indicated by the highest enum value (some examples are: SSH, HTTP, DNS, SMB).
+
+_`Examples`
+===========
+
+Enums
+~~~~~~
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:130]: nit -- extends past text.

File: doc/devguide/extending/app-layer/transactions.rst [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+    enum {
+        FTP_STATE_IN_PROGRESS,
+        FTP_STATE_PORT_DONE,
+        FTP_STATE_FINISHED,
+    };
+
+
+API Callbacks
+~~~~~~~~~~~~~~
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:155]: nit -- too long.

File: doc/devguide/extending/app-layer/transactions.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    AppLayerParserRegisterStateProgressCompletionStatus(
+        ALPROTO_FTP, FTP_STATE_FINISHED, FTP_STATE_FINISHED);
+
+_`Common words and abbreviations`
+=================================
+
+- al, applayer: application layer
+- alproto: application layer protocol
+- alstate: application layer state
+- engine: refers to Suricata core
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:218]: Suggest "Suricata core detection logic" ?

File: doc/devguide/extending/app-layer/transactions.rst [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                    }
+                }
+             }
+        .
+        .
+        .
+    }
+
+Rule Matching
+~~~~~~~~~~~~~~
```
Comment (jlucovsky) [inline doc/devguide/extending/app-layer/transactions.rst:68]: nit: extends past text.

**Review by jlucovsky — COMMENTED** [review]
Looks good @jufajardini ! Minor comments

**Comment by jufajardini** [comment]
Changes incorporated in: https://github.com/OISF/suricata/pull/6220

**Pull request closed without merging**